pytorch_lightning/strategies/collaborative.py

```diff
@@ -0,0 +1,504 @@
+            if self._batch_size is None:
+                try:
+                    self._batch_size = extract_batch_size(batch)
+                    log.info(f"Found per machine batch size automatically from the batch: {self._batch_size}")
+                except Exception as e:
+                    raise MisconfigurationException(
+                        "We tried to infer the batch size from the first batch of data are were unable to. "
+                        "Please provide the batch size to the Strategy (CollaborativeStrategy"
+                        f"exception raised: {e}"
+                    )
+            self._initialize_hivemind()
+
+    def reduce(self, tensor: Union[Any, torch.Tensor], *args: Any, **kwargs: Any) -> Union[Any, torch.Tensor]:
+        return tensor
+
+    def all_gather(self, tensor: torch.Tensor, group: Optional[Any] = None, sync_grads: bool = False) -> torch.Tensor:
+        return tensor
+
+    def model_to_device(self) -> None:
+        assert self.model is not None
+        self.model.to(self.root_device)
+
+    def barrier(self, *args: Any, **kwargs: Any) -> None:
+        pass
+
+    def broadcast(self, obj: TBroadcast, src: int = 0) -> TBroadcast:
+        return obj
+
+    def teardown(self) -> None:
+        super().teardown()
+        if self.root_device.type == "cuda" and self.lightning_module is not None:
+            # GPU teardown
+            self.lightning_module.cpu()
+            # clean up memory
+            torch.cuda.empty_cache()
+        if self._opt:
+            self._opt.shutdown()
+        log.info("Shutting down hivemind DHT.")
+        self.dht.shutdown()
+
+
+class HiveMindScheduler:
+    """Wrapper for schedulers to prevent Lightning from stepping the scheduler too soon.
+
+    This code ensures that we only step when the HiveMind optimizer reaches the global step.
+    """
+
+    def __init__(self, optimizer: "hivemind.Optimizer", scheduler: LRSchedulerTypeUnion) -> None:
+        # copy most of the `Scheduler` methods into this instance. `__del__` is skipped in case the scheduler has
+        # implemented custom logic which we would not want to call on destruction of the `HiveMindScheduler`
+        self.__dict__ = {k: v for k, v in scheduler.__dict__.items() if k not in ("step", "__del__")}
+
+        self.optimizer = optimizer
+        self.scheduler = scheduler
+        self.current_step = -1
+
+    def step(self, epoch: Optional[int] = None) -> None:
+        while self.current_step < self.optimizer.local_epoch:
+            self.scheduler.step()
+            self.current_step += 1
+
+    def load_state_dict(self, state_dict: Dict) -> None:
+        self.scheduler.load_state_dict(state_dict)
+
+    def state_dict(self) -> Dict:
+        return self.scheduler.state_dict()
+
+
+class DHTManager:
+    ENDPOINT_ENV: str = "ENDPOINT"
+    PEER_ENDPOINT_ENV: str = "PEER_ENDPOINT"
+    INITIAL_PEERS_ENV: str = "INITIAL_PEERS"
+    HOST_ENV: str = "HOST"
+    PORT_ENV: str = "PORT"
```

Comment:
We usually add `PL_` prefix to the env variables used by PyTorch Lightning internally.
Should we do that here as well to avoid potential conflicts?